### ISSUE TITLE
Automatic decoding in the `http` source

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#4217]
The `http` source offers an `encoding` option that default to `"text"`. This tells the source how to decode the data. This is uncessary since the incoming request should have a `Content-Type` header. We should decode the data based on the provided `Content-Type` header and error when that header is not present.

**Comment by jszwedko** [comment]
We may want to draw a distinction between `Content-Type` and `Content-Encoding` for HTTP based sources.

For `application/json` it'd be nice to allow the user to specify how to parse it for cases where they don't control the publisher.

**Comment by binarylogic** [comment]
> For application/json it'd be nice to allow the user to specify how to parse it for cases where they don't control the publisher (for example, maybe their request structure looks like { "events": [ ... ]}.

This sounds like a job for the remap language, right?

**Comment by jszwedko** [comment]
:+1: I think `remap` could support it. It seems like it would conflict with:

> Only an array of objects is allowed, single root-level objects are not allowed.

though, no? Would users use ndjson in that case?

**Comment by binarylogic** [comment]
True. My intent was to guide users away from firing one event per request due to performance; something I've seen users do with other tools 😢 . Your use case violates that though.
